appsec/src/extension/backtrace.c

```diff
@@ -0,0 +1,280 @@
+// Unless explicitly stated otherwise all files in this repository are
+// dual-licensed under the Apache-2.0 License or BSD-3-Clause License.
+//
+// This product includes software developed at Datadog
+// (https://www.datadoghq.com/). Copyright 2021 Datadog, Inc.
+#include "backtrace.h"
+#include "configuration.h"
+#include "ddtrace.h"
+#include "logging.h"
+#include "php_compat.h"
+#include "php_objects.h"
+#include "string_helpers.h"
+
+static const int NO_LIMIT = 0;
+static const double STACK_DEFAULT_TOP_PERCENTAGE = 0.25;
+
+static zend_string *_frames_key;
+static zend_string *_language_key;
+static zend_string *_php_value;
+static zend_string *_exploit_key;
+static zend_string *_dd_stack_key;
+static zend_string *_frame_line;
+static zend_string *_frame_function;
+static zend_string *_frame_file;
+static zend_string *_id_key;
+
+bool php_backtrace_frame_to_datadog_backtrace_frame( // NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+    zval *php_backtrace_frame, zval *datadog_backtrace_frame, zend_ulong index)
+{
+    if (Z_TYPE_P(php_backtrace_frame) != IS_ARRAY) {
+        return false;
+    }
+    HashTable *frame = Z_ARRVAL_P(php_backtrace_frame);
+    zval *line = zend_hash_str_find(frame, "line", sizeof("line") - 1);
+    zval *function =
+        zend_hash_str_find(frame, "function", sizeof("function") - 1);
+    zval *file = zend_hash_str_find(frame, "file", sizeof("file") - 1);
+    zval id;
+    ZVAL_LONG(&id, index);
+
+#ifdef TESTING
+    // In order to be able to test full path encoded everywhere lets set
+    // only the file name without path
+    char *file_name = strrchr(Z_STRVAL_P(file), '/');
```

Comment:
This can return `NULL` as well if the `/` isn't present.

Reply:
This is just for testing purposes